### ISSUE TITLE
Airsonic API is returning different results from folders to which the user does not have access.

### ISSUE THREAD

**Issue by SteveDinn** [airsonic-advanced/airsonic-advanced#364]
## Problem description
There are a few cases where I've noticed that the airsonic API is returning different results from the web interface. I believe this is because it's querying from all folders, not just the ones to which the user has access.

### Steps to reproduce

I'm sure there may be a simpler way to reproduce this, but I'll detail the way my music is organized: I have music folders set up for me, my wife, and my kids. I have a master folder for all our music, and then I hard link (**not** symlink, I use this command: "cp -arl") the music from the master folder into each of our folders, so that we don't have to listen to each others' music, and there isn't any duplication of files.

I then create separate media folders for each of us in airsonic-advanced and each user gets access only to their own folder. Via the airsonic-advanced website, this works well, but if more than one of us has a folder hard-linked into our music library, we'll see duplicates of tracks. I realize these tracks *are* actually duplicated, but a user should only see the tracks that are under folders that they're allowed to see.

This same situation also seems to prevent those artists that are duplicated across libraries from coming back in the artist listings, similarly for artists in the artists listing, does not show any albums under the artist listing, and prevents both those artists and albums from appearing in search results.

Let me know if I should be more verbose here. I can test if necessary.

## System information

 * **Airsonic version**: 11.0.0-SNAPSHOT.20200717173203 – July 17, 2020 at 2:32:03 PM ADT
 * **Operating system**: Official docker image run on Ubuntu 16.04
 * **Java version**: Official docker image version
 * **Proxy server**: *None
 * **Client**: current Chrome stable
 * **Language**: English

## Additional notes

*Include any extra notes here. Otherwise you may remove this section.*

**Comment by SteveDinn** [comment]
To be more specific here:
* The same album is in two top-level library folders.
* These library folders are mutually exclusive to users.
* Airsonic API returns things from library folders that the user does not have access to.
* Despite being returned, Airsonic-Advanced won't let the user play those tracks.
* If the same album is present in multiple library folders, the albums are duplicated. This would be the expected behavior if the user had access to both folders.

I get that this is kind of crossing the boundaries between folder browsing and tagged browsing, but I feel like there should be a distinction when it comes to a user's access to those top-level library folders.

**Comment by randomnicode** [comment]
Trying to get clarity on the issue. Let's see if these assumptions are correct. For the sake of clarity let's just talk about one album A1.

- You have an album A1 in parent folder P1 that airsonic _does NOT_ have access to: _P1/A1_
- You have two folders U1 and U2 each of which airsonic _does_ have access to, and is assigned mutually exclusive users permissions. User1 has access to U1 but not U2, and User2 has access to U2 but not U1.
- You have a hardlink in U1 and U2 pointing to P1/A1: _U1/A1a -> P1/A1_ and _U2/A1b -> P1/A1_

Okay, Airsonic sees each music file as its own now. There are two albums effectively as far as Airsonic is concerned, A1a and A1b
Assuming this structure holds, what exactly is the issue?

> Airsonic API returns things from library folders that the user does not have access to.

Which API?
Search? The API for the searchbox you use at the top of the web UI should filter out all folders a user does not have access to.

https://github.com/airsonic-advanced/airsonic-advanced/blob/278e87b37be364588fbe2208875bf4d1f92aad72/airsonic-main/src/main/java/org/airsonic/player/controller/SearchController.java#L82-L92

Notice that we retrieve the folders a user has access to and search only those folders.

Where exactly are you seeing duplicates?

**Comment by SteveDinn** [comment]
I've reproduced this in as simple a manner as I could.  Here's my docker-compose.yml:
```
  airsonic-test:
    image: airsonicadvanced/airsonic-advanced:latest
    container_name: airsonic-test
    hostname: airsonic-test
    restart: "no"
    ports:
      - 88:4040
    volumes:
      - ./airsonic-test/data:/var/airsonic
      - ./airsonic-test/playlists:/var/playlists
      - ./airsonic-test/podcasts:/var/podcasts
      - ./airsonic-test/music:/var/music
    environment:
      - TZ=America/Halifax
```

Create 2 music library folders, `/var/music/1` and `/var/music/2`.
Each folder has one artist (and their albums) unique to it, and one artist (and albums) that they both have.
Create 2 users and give each user access to one of the folders

![image](https://user-images.githubusercontent.com/3911507/93338547-d7a7ac80-f800-11ea-9d12-bf04b150e037.png)

Scan folders, and login using an app that uses the Airsonic API.

Here is the content of those folders. Note that they both contain the same set of albums for the band "Metric". These are literally the same files copied over into both folders.
```
root@orion /srv/docker/airsonic-test/music # dir 1
total 8
drwxrwsr-x 5 subsonic media 4096 Aug 17 00:29 Metric/
drwxrwsr-x 5 subsonic media 4096 Jun 14  2019 Smashing Pumpkins/
root@orion /srv/docker/airsonic-test/music # dir 2
total 8
drwxrwsr-x 6 subsonic media 4096 Aug 26 16:29 Katy Perry/
drwxrwsr-x 5 subsonic media 4096 Aug 17 00:29 Metric/
root@orion /srv/docker/airsonic-test/music #
```
Through the native web UI, everything is as expected. Metric albums show up once, and all songs show up once per album.

![image](https://user-images.githubusercontent.com/3911507/93345216-ee520180-f808-11ea-9ae3-cfe1e089db70.png)

I've been using [@tamland's airsonic frontend](https://github.com/tamland/airsonic-frontend) as a client, and this uses the Airsonic API. When I log in with this client, as one of the non-admin users, I can see each album only once, which is good, but the albums have duplicated tracks. I can only assume that this is because the album is present in more than one library folder. In fact, when I add /var/music/3 with a third copy of that album, I see the tracks in triplicate despite the user only having access to one of those root library folders.

![image](https://user-images.githubusercontent.com/3911507/93345334-13467480-f809-11ea-996e-b145fcb461a5.png)

The search also looked messed up to me previously (only showing tracks, no albums), but for whatever reason, I am unable to reproduce that reliably, so forget about that for now. Hopefully I've covered all the bases here. I know this is a weird one.

**Comment by SteveDinn** [comment]
There is something weird going on between the native airsonic web client and tamland's web client. This might deserve to be separate bug.
* Login as one user in the native web client
* In another tab in the same browser, login as a different user in airsonic-frontend
* Play a song in airsonic-frontend
* See that airsonic log shows that the _native web client user_ has played the song.
* This also lets users play songs to which to they do not have access if they were displayed via the original bug.

If there is no login to the native web UI, then the log shows the correct user and does correct access checking.

I believe this has to be a bug in Airsonic-Advanced, because the third party client has no knowledge of my 'admin' user, yet is playing songs as them!

**Comment by SteveDinn** [comment]
In one of the last few docker images, this changed slightly. I still see multiple tracks from multiple folders when I should only see one, but I can no longer play the ones that aren't under folders I have access to.
